synapse/logging/filter.py

```diff
@@ -0,0 +1,33 @@
+# -*- coding: utf-8 -*-
+# Copyright 2020 The Matrix.org Foundation C.I.C.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import logging
+
+from typing_extensions import Literal
+
+
+class MetadataFilter(logging.Filter):
+    """Logging filter that adds constant values to each record.
+
+    Args:
+        metadata: Key-value pairs to add to each record.
+    """
+
+    def __init__(self, metadata: dict):
+        self._metadata = metadata
+
+    def filter(self, record: logging.LogRecord) -> Literal[True]:
+        for key, value in self._metadata.items():
+            setattr(record, key, value)
```

Comment:
Attaching fields to an object that we don't own feels, dangerous? There'd be nothing stopping the stdlib from making the `LogRecord` use `__slots__`?

Reply:
Ah, but turns out that `extra` kwarg of `logger.debug` and co. populate the `__dict__` of the LogRecord

Reply:
Yeah, this is essentially a supported thing to do and examples in the standard library docs do this.

Reply:
The [documentation about adding info in filters](https://docs.python.org/3/howto/logging-cookbook.html#filters-contextual).